### ISSUE TITLE
[ARM, debug] `test_topology_remove_garbage_group0` failed

### ISSUE THREAD

**Issue by Annamikhlin** [scylladb/scylladb#16306]
`topology_custom.test_topology_remove_garbage_group0` failed on ARM, debug 
ERROR:
```
=================================== FAILURES ===================================
______________________ test_remove_garbage_group0_members ______________________

self = <test.pylib.manager_client.ManagerClient object at 0xffff5b26a150>
replace_cfg = None, cmdline = None
config = {'enable_user_defined_functions': False, 'experimental_features': []}
property_file = None, start = True, expected_error = None

    async def server_add(self, replace_cfg: Optional[ReplaceConfig] = None,
                         cmdline: Optional[List[str]] = None,
                         config: Optional[dict[str, Any]] = None,
                         property_file: Optional[dict[str, Any]] = None,
                         start: bool = True,
                         expected_error: Optional[str] = None) -> ServerInfo:
        """Add a new server"""
        try:
            data = self._create_server_add_data(replace_cfg, cmdline, config, property_file, start, expected_error)
    
            # If we replace, we should wait until other nodes see the node being
            # replaced as dead because the replace operation can be rejected if
            # the node being replaced is considered alive. However, we sometimes
            # do not want to wait, for example, when we test that replace fails
            # as expected. Therefore, we make waiting optional and default.
            if replace_cfg and replace_cfg.wait_replaced_dead:
                replaced_ip = await self.get_host_ip(replace_cfg.replaced_id)
                await self.others_not_see_server(replaced_ip)
    
>           server_info = await self.client.put_json("/cluster/addserver", data, response_type="json",
                                                     timeout=ScyllaServer.TOPOLOGY_TIMEOUT)

test/pylib/manager_client.py:215: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <test.pylib.rest_client.UnixRESTClient object at 0xffff5b229c50>
resource_uri = '/cluster/addserver'
data = {'config': {'enable_user_defined_functions': False, 'experimental_features': []}, 'start': True}
host = None, port = None, params = None, response_type = 'json', timeout = 1000

    async def put_json(self, resource_uri: str, data: Optional[Mapping] = None, host: Optional[str] = None,
                       port: Optional[int] = None, params: Optional[dict[str, str]] = None,
                       response_type: Optional[str] = None, timeout: Optional[float] = None) -> Any:
>       ret = await self._fetch("PUT", resource_uri, response_type = response_type, host = host,
                                port = port, params = params, json = data, timeout = timeout)

test/pylib/rest_client.py:104: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <test.pylib.rest_client.UnixRESTClient object at 0xffff5b229c50>
method = 'PUT', resource = '/cluster/addserver', response_type = 'json'
host = None, port = None, params = None
json = {'config': {'enable_user_defined_functions': False, 'experimental_features': []}, 'start': True}
timeout = 1000

    async def _fetch(self, method: str, resource: str, response_type: Optional[str] = None,
                     host: Optional[str] = None, port: Optional[int] = None,
                     params: Optional[Mapping[str, str]] = None,
                     json: Optional[Mapping] = None, timeout: Optional[float] = None) -> Any:
        # Can raise exception. See https://docs.aiohttp.org/en/latest/web_exceptions.html
        assert method in ["GET", "POST", "PUT", "DELETE"], f"Invalid HTTP request method {method}"
        assert response_type is None or response_type in ["text", "json"], \
                f"Invalid response type requested {response_type} (expected 'text' or 'json')"
        # Build the URI
        port = port if port else self.default_port if hasattr(self, "default_port") else None
        port_str = f":{port}" if port else ""
        assert host is not None or hasattr(self, "default_host"), "_fetch: missing host for " \
                "{method} {resource}"
        host_str = host if host is not None else self.default_host
        uri = self.uri_scheme + "://" + host_str + port_str + resource
        logging.debug(f"RESTClient fetching {method} {uri}")
    
        client_timeout = ClientTimeout(total = timeout if timeout is not None else 300)
>       async with request(method, uri,
                           connector = self.connector if hasattr(self, "connector") else None,
                           params = params, json = json, timeout = client_timeout) as resp:

test/pylib/rest_client.py:64: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <aiohttp.client._SessionRequestContextManager object at 0xffff5a116f00>

    async def __aenter__(self) -> ClientResponse:
        try:
>           self._resp = await self._coro

/usr/lib64/python3.11/site-packages/aiohttp/client.py:1190: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <aiohttp.client.ClientSession object at 0xffff5bc5acd0>, method = 'PUT'
str_or_url = 'http+unix://api/cluster/addserver'

    async def _request(
        self,
        method: str,
        str_or_url: StrOrURL,
        *,
        params: Optional[Mapping[str, str]] = None,
        data: Any = None,
        json: Any = None,
        cookies: Optional[LooseCookies] = None,
        headers: Optional[LooseHeaders] = None,
        skip_auto_headers: Optional[Iterable[str]] = None,
        auth: Optional[BasicAuth] = None,
        allow_redirects: bool = True,
        max_redirects: int = 10,
        compress: Optional[str] = None,
        chunked: Optional[bool] = None,
        expect100: bool = False,
        raise_for_status: Optional[bool] = None,
        read_until_eof: bool = True,
        proxy: Optional[StrOrURL] = None,
        proxy_auth: Optional[BasicAuth] = None,
        timeout: Union[ClientTimeout, object] = sentinel,
        verify_ssl: Optional[bool] = None,
        fingerprint: Optional[bytes] = None,
        ssl_context: Optional[SSLContext] = None,
        ssl: Optional[Union[SSLContext, bool, Fingerprint]] = None,
        proxy_headers: Optional[LooseHeaders] = None,
        trace_request_ctx: Optional[SimpleNamespace] = None,
        read_bufsize: Optional[int] = None,
    ) -> ClientResponse:
    
        # NOTE: timeout clamps existing connect and read timeouts.  We cannot
        # set the default to None because we need to detect if the user wants
        # to use the existing timeouts by setting timeout to None.
    
        if self.closed:
            raise RuntimeError("Session is closed")
    
        ssl = _merge_ssl_params(ssl, verify_ssl, ssl_context, fingerprint)
    
        if data is not None and json is not None:
            raise ValueError(
                "data and json parameters can not be used at the same time"
            )
        elif json is not None:
            data = payload.JsonPayload(json, dumps=self._json_serialize)
    
        if not isinstance(chunked, bool) and chunked is not None:
            warnings.
...[truncated 2442 chars]...
rt explicit
                    # Authorization header with auth argument
                    if (
                        headers is not None
                        and auth is not None
                        and hdrs.AUTHORIZATION in headers
                    ):
                        raise ValueError(
                            "Cannot combine AUTHORIZATION header "
                            "with AUTH argument or credentials "
                            "encoded in URL"
                        )
    
                    all_cookies = self._cookie_jar.filter_cookies(url)
    
                    if cookies is not None:
                        tmp_cookie_jar = CookieJar()
                        tmp_cookie_jar.update_cookies(cookies)
                        req_cookies = tmp_cookie_jar.filter_cookies(url)
                        if req_cookies:
                            all_cookies.load(req_cookies)
    
                    if proxy is not None:
                        proxy = URL(proxy)
                    elif self._trust_env:
                        with suppress(LookupError):
                            proxy, proxy_auth = get_env_proxy_for_url(url)
    
                    req = self._request_class(
                        method,
                        url,
                        params=params,
                        headers=headers,
                        skip_auto_headers=skip_headers,
                        data=data,
                        cookies=all_cookies,
                        auth=auth,
                        version=version,
                        compress=compress,
                        chunked=chunked,
                        expect100=expect100,
                        loop=self._loop,
                        response_class=self._response_class,
                        proxy=proxy,
                        proxy_auth=proxy_auth,
                        timer=timer,
                        session=self,
                        ssl=ssl,
                        proxy_headers=proxy_headers,
                        traces=traces,
                    )
    
                    # connection timeout
                    try:
                        async with ceil_timeout(real_timeout.connect):
                            assert self._connector is not None
                            conn = await self._connector.connect(
                                req, traces=traces, timeout=real_timeout
                            )
                    except asyncio.TimeoutError as exc:
                        raise ServerTimeoutError(
                            "Connection timeout " "to host {}".format(url)
                        ) from exc
    
                    assert conn.transport is not None
    
                    assert conn.protocol is not None
                    conn.protocol.set_response_params(
                        timer=timer,
                        skip_payload=method.upper() == "HEAD",
                        read_until_eof=read_until_eof,
                        auto_decompress=self._auto_decompress,
                        read_timeout=real_timeout.sock_read,
                        read_bufsize=read_bufsize,
                    )
    
                    try:
                        try:
                            resp = await req.send(conn)
                            try:
>                               await resp.start(conn)

/usr/lib64/python3.11/site-packages/aiohttp/client.py:561: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ClientResponse(http+unix://api/cluster/addserver) [None None]>
None

connection = Connection<ConnectionKey(host='api', port=None, is_ssl=False, ssl=None, proxy=None, proxy_auth=None, proxy_headers_hash=None)>

    async def start(self, connection: "Connection") -> "ClientResponse":
        """Start response processing."""
        self._closed = False
        self._protocol = connection.protocol
        self._connection = connection
    
>       with self._timer:

/usr/lib64/python3.11/site-packages/aiohttp/client_reqrep.py:894: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <aiohttp.helpers.TimerContext object at 0xffff5a113490>
exc_type = <class 'asyncio.exceptions.CancelledError'>
exc_val = CancelledError(), exc_tb = <traceback object at 0xffff5a0ce3c0>

    def __exit__(
        self,
        exc_type: Optional[Type[BaseException]],
        exc_val: Optional[BaseException],
        exc_tb: Optional[TracebackType],
    ) -> Optional[bool]:
        if self._tasks:
            self._tasks.pop()
    
        if exc_type is asyncio.CancelledError and self._cancelled:
>           raise asyncio.TimeoutError from None
E           TimeoutError

/usr/lib64/python3.11/site-packages/aiohttp/helpers.py:721: TimeoutError

The above exception was the direct cause of the following exception:

manager = <test.pylib.manager_client.ManagerClient object at 0xffff5b26a150>

    async def test_remove_garbage_group0_members(manager: ManagerClient):
        """
        Verify that failing to leave group 0 or remove a node from group 0 in removenode/decommission
        can be handled by executing removenode (which should clear the 'garbage' group 0 member),
        even though the node is no longer a token ring member. Does not apply to Raft-topology mode.
        """
        # 4 servers, one dead
        cfg = {'enable_user_defined_functions': False,
               'experimental_features': list[str]()}
>       servers = [await manager.server_add(config=cfg) for _ in range(4)]

test/topology_custom/test_topology_remove_garbage_group0.py:29: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/topology_custom/test_topology_remove_garbage_group0.py:29: in <listcomp>
    servers = [await manager.server_add(config=cfg) for _ in range(4)]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <test.pylib.manager_client.ManagerClient object at 0xffff5b26a150>
replace_cfg = None, cmdline = None
config = {'enable_user_defined_functions': False, 'experimental_features': []}
property_file = None, start = True, expected_error = None

    async def server_add(self, replace_cfg: Optional[ReplaceConfig] = None,
                         cmdline: Optional[List[str]] = None,
                         config: Optional[dict[str, Any]] = None,
                         property_file: Optional[dict[str, Any]] = None,
                         start: bool = True,
                         expected_error: Optional[str] = None) -> ServerInfo:
        """Add a new server"""
        try:
            data = self._create_server_add_data(replace_cfg, cmdline, config, property_file, start, expected_error)
    
            # If we replace, we should wait until other nodes see the node being
            # replaced as dead because the replace operation can be rejected if
            # the node being replaced is considered alive. However, we sometimes
            # do not want to wait, for example, when we test that replace fails
            # as expected. Therefore, we make waiting optional and default.
            if replace_cfg and replace_cfg.wait_replaced_dead:
                replaced_ip = await self.get_host_ip(replace_cfg.replaced_id)
                await self.others_not_see_server(replaced_ip)
    
            server_info = await self.client.put_json("/cluster/addserver", data, response_type="json",
                                                     timeout=ScyllaServer.TOPOLOGY_TIMEOUT)
        except Exception as exc:
>           raise Exception("Failed to add server") from exc
E           Exception: Failed to add server
```
link to the failure: https://jenkins.scylladb.com/job/scylla-master/job/build/1716/testReport/(root)/non-boost%20tests/Build__Test__Publish___buildTestPublishArm___topology_custom_test_topology_remove_garbage_group0_debug_1/
link to the full log: https://jenkins.scylladb.com/job/scylla-master/job/build/1716/artifact/testlog/aarch64/debug/topology_custom.test_topology_remove_garbage_group0.1.log

**Comment by avikivity** [comment]
Seen again: https://jenkins.scylladb.com/view/nexts/job/scylla-master/job/next/7327/testReport/junit/(root)/test_topology_remove_garbage_group0/Build___x86___Unit_Tests_x86___test_remove_garbage_group0_members/

**Comment by kbr-scylla** [comment]
> Seen again: https://jenkins.scylladb.com/view/nexts/job/scylla-master/job/next/7327/testReport/junit/(root)/test_topology_remove_garbage_group0/Build___x86___Unit_Tests_x86___test_remove_garbage_group0_members/

Different error
```
    
>       group0_members = await get_current_group0_config(manager, servers[3])

test/topology_custom/test_topology_remove_garbage_group0.py:109: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/topology/util.py:67: in get_current_group0_config
    host = (await wait_for_cql_and_get_hosts(manager.cql, [srv], time.time() + 60))[0]
test/pylib/util.py:81: in wait_for_cql_and_get_hosts
    hosts = await wait_for(
test/pylib/util.py:53: in wait_for
    before_retry()
test/pylib/util.py:84: in <lambda>
    before_retry=lambda: cql.cluster.refresh_nodes(force_token_rebuild=True),
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   cassandra.DriverException: Node list was not refreshed. See log for details.
```

(and already fixed -- ec1f574b3a8afc4ee40af4fe4e19c90004706d48)

**Comment by kbr-scylla** [comment]
The logs from original post are missing, closing.
